test/unit/org/apache/cassandra/gms/GossiperTest.java

```diff
@@ -414,11 +417,95 @@ public void testNotFireDuplicatedNotificationsWithUpdateContainsOldAndNewState()
+            proposedRemoteState.addApplicationState(ApplicationState.STATUS_WITH_PORT, valueFactory.normal(Collections.singletonList(token)));
+            Gossiper.instance.applyStateLocally(ImmutableMap.of(remoteHostAddress1, proposedRemoteState));
+
+            // we still have only 1 endpoint with 2 transitions
+            assertThat(Gossiper.instance.gossipStateTransitionMap.size()).isEqualTo(1);
+            assertThat(Gossiper.instance.gossipStateTransitionMap.get(remoteHostAddress1).size()).isEqualTo(2);
+
+            // STEP 3. update the heartbeat state
+            proposedRemoteHeartBeat = new HeartBeatState(initialRemoteHeartBeat.getGeneration() + Gossiper.MAX_GENERATION_DIFFERENCE + 1);
+            proposedRemoteState = new EndpointState(proposedRemoteHeartBeat);
+            Gossiper.instance.applyStateLocally(ImmutableMap.of(remoteHostAddress1, proposedRemoteState));
+
+            // we still have only 1 endpoint
+            assertThat(Gossiper.instance.gossipStateTransitionMap.size()).isEqualTo(1);
+            assertThat(Gossiper.instance.gossipStateTransitionMap.get(remoteHostAddress1).size()).isEqualTo(2);
+            assertThat(Gossiper.instance.gossipStateTransitionMap.get(remoteHostAddress1).peek().status).contains(":NORMAL," + token);
+
+            InetAddressAndPort remoteHostAddress2 = hosts.get(2);
+            initialRemoteState = Gossiper.instance.getEndpointStateForEndpoint(remoteHostAddress2);
+            initialRemoteHeartBeat = initialRemoteState.getHeartBeatState();
+            //Util.createInitialRing should have initialized remoteHost's HeartBeatState's generation to 1 for
+            //the second host
+            assertThat(initialRemoteHeartBeat.getGeneration()).isEqualTo(1);
+
+            proposedRemoteHeartBeat = new HeartBeatState(initialRemoteHeartBeat.getGeneration());
+            proposedRemoteState = new EndpointState(proposedRemoteHeartBeat);
+            token = DatabaseDescriptor.getPartitioner().getRandomToken();
+            proposedRemoteState.addApplicationState(ApplicationState.STATUS, valueFactory.normal(Collections.singletonList(token)));
+
+            Gossiper.instance.applyStateLocally(ImmutableMap.of(remoteHostAddress2, proposedRemoteState));
+
+            // now we have 2 endpoints in the gossipStateTransitionMap
+            assertThat(Gossiper.instance.gossipStateTransitionMap.size()).isEqualTo(2);
+            assertThat(Gossiper.instance.gossipStateTransitionMap.get(remoteHostAddress1).size()).isEqualTo(2);
+            assertThat(Gossiper.instance.gossipStateTransitionMap.get(remoteHostAddress2).size()).isEqualTo(1);
+
+            // the list of recent gossip state transtitions includes 2 transitions from host 1 and 1 transition
+            // from host 2
+            List<GossipStateTransition> gossipStateTransitions = Gossiper.instance.recentGossipStateTransitions();
+            assertThat(gossipStateTransitions.size()).isEqualTo(3);
```

Comment:
Seems like removing a endpoint from membership is not tested,  we can add below lines to test that code path as well

```
           Gossiper.instance.replacedEndpoint(remoteHostAddress1);
            List<GossipStateTransition> gossipStateTransitionsAfterReplace = Gossiper.instance.recentGossipStateTransitions();
            assertThat(gossipStateTransitionsAfterReplace.size()).isEqualTo(1);
```